framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,454 @@
 
-    @Pure // because the type field is final
-    public AnnotatedTypeMirror getType() {
-        return type;
+        validateSet(this.getAnnotations(), this.getUnderlyingType());
     }
 
-    @Override
-    public V leastUpperBound(/*@Nullable*/ V other) {
-        if (other == null) {
-            @SuppressWarnings("unchecked")
-            V v = (V) this;
-            return v;
+    private boolean validateSet(Set<AnnotationMirror> annos, TypeMirror typeMirror) {
+        if (canBeMissingAnnotations(typeMirror)) {
+            return true;
         }
-        AnnotatedTypeMirror otherType = other.getType();
-        AnnotatedTypeMirror type = getType();
 
-        GenericAnnotatedTypeFactory<V, ?, ?, ?> factory = analysis.getTypeFactory();
-        ProcessingEnvironment processingEnv = factory.getProcessingEnv();
-        AnnotatedTypeMirror lubAnnotatedType =
-                AnnotatedTypes.leastUpperBound(processingEnv, factory, type, otherType);
+        Set<AnnotationMirror> missingHierarchy = null;
+        for (AnnotationMirror top : hierarchy.getTopAnnotations()) {
+            AnnotationMirror anno = hierarchy.findAnnotationInHierarchy(annos, top);
+            if (anno == null) {
+                if (missingHierarchy == null) {
+                    missingHierarchy = AnnotationUtils.createAnnotationSet();
+                }
+                missingHierarchy.add(top);
+            }
+        }
+
+        if (missingHierarchy != null) {
+            String missing = PluginUtil.join(", ", missingHierarchy);
+            ErrorReporter.errorAbort(
+                    "CFAbstractValue: missing annotation in the following hierarchies: " + missing);
+            return false;
+        }
+        return true;
+    }
 
-        return analysis.createAbstractValue(lubAnnotatedType);
+    /**
+     * Returns whether or not the set of annotations can be missing an annotation for any hierarchy
+     * @return whether or not the set of annotations can be missing an annotation
+     */
+    public boolean canBeMissingAnnotations() {
+        return canBeMissingAnnotations(underlyingType);
     }
 
-    private static void copyArrayComponentAnnotations(
-            AnnotatedArrayType source, AnnotatedArrayType dest) {
-        AnnotatedTypeMirror destComp = dest.getComponentType();
-        AnnotatedTypeMirror sourceComp = source.getComponentType();
-        destComp.addAnnotations(sourceComp.getAnnotations());
-        if (sourceComp instanceof AnnotatedArrayType) {
-            assert dest instanceof AnnotatedArrayType;
-            copyArrayComponentAnnotations(
-                    (AnnotatedArrayType) sourceComp, (AnnotatedArrayType) destComp);
+    private boolean canBeMissingAnnotations(TypeMirror typeMirror) {
+        if (typeMirror == null) {
+            return false;
+        }
+        if (typeMirror.getKind() == TypeKind.WILDCARD) {
+            return canBeMissingAnnotations(((WildcardType) typeMirror).getExtendsBound());
```

Comment:
Why different logic for wildcards and type variables? Why not the same recursive call in both cases?